pd/src/consensus/worker.rs

```diff
@@ -247,164 +241,235 @@ impl Worker {
+    async fn end_epoch(&mut self) -> Result<()> {
+        let reader = self.state.private_reader();
+
+        let pending_block = self
+            .pending_block
+            .as_mut()
+            .expect("pending block must be Some in EndBlock");
 
-            // steps (foreach validator):
-            // - get the total token supply for the validator's delegation tokens
-            // - process the updates to the token supply:
-            //   - collect all delegations occurring in previous epoch and apply them (adds to supply);
-            //   - collect all undelegations started in previous epoch and apply them (reduces supply);
-            // - feed the updated (current) token supply into current_rates.voting_power()
-            // - persist both the current voting power and the current supply
-            //
+        let height = pending_block
+            .height
+            .expect("height must already have been set");
+
+        // We've finished processing the last block of `epoch`, so we've
+        // crossed the epoch boundary, and (prev | current | next) are:
+        let prev_epoch = pending_block
+            .epoch
+            .clone()
+            .expect("epoch must already have been set");
+        let current_epoch = prev_epoch.next();
+        let next_epoch = current_epoch.next();
+
+        tracing::info!(
+            ?height,
+            ?prev_epoch,
+            ?current_epoch,
+            ?next_epoch,
+            "crossed epoch boundary, processing rate updates"
+        );
+        metrics::increment_counter!("epoch");
 
-            /// FIXME: set this less arbitrarily, and allow this to be set per-epoch
-            /// 3bps -> 11% return over 365 epochs, why not
-            const BASE_REWARD_RATE: u64 = 3_0000;
+        // Find all the validators which were *not* slashed in this block
+        let well_behaved_validators = pending_block
+            .validator_state_changes
+            .iter()
+            // THIS IS A LOAD-BEARING NEGATION: we want all validators which are *NOT* slashed
+            .filter(|p| !matches!(p.1, ValidatorState::Slashed))
+            .map(|p| p.0.clone()) // get the validator ID
+            .collect::<Vec<_>>();
+
+        // Process unbonding notes and nullifiers for this epoch
+        let (mut unbonding_notes, mut unbonding_nullifiers) = (
+            reader.quarantined_notes(Some(height), Some(well_behaved_validators.iter())),
+            reader.quarantined_nullifiers(Some(height), Some(well_behaved_validators.iter())),
+        );
+        while let Some(result) = unbonding_notes.next().await {
+            let (_, commitment, data) = result?;
+            pending_block.add_note(commitment, data);
+        }
+        while let Some(result) = unbonding_nullifiers.next().await {
+            pending_block.unbonding_nullifiers.insert(result?.1); // insert the nullifier
+        }
+        drop(unbonding_notes);
+        drop(unbonding_nullifiers);
+
```

Comment:
The rest of this file's changes are not things I've touched; this is re-arrangement noise due to the rebase and the creation of the separated `end_epoch` function.